### PR TITLE
chore(types): generate our own d.ts file from api.md

### PR DESCRIPTION
**Pull request by JoelEinbinder**

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1032,7 +1032,7 @@ Shortcut for [page.mainFrame().$$(selector)](#frameselector-1).
 
 #### page.$$eval(selector, pageFunction[, ...args])
 - `selector` <[string]> A [selector] to query page for
-- `pageFunction` <[function]> Function to be evaluated in browser context
+- `pageFunction` <[function]\([Array]<[Element]>\)> Function to be evaluated in browser context
```
Comment (aslushnikov) [inline docs/api.md:1038]: escaping parenthesis is very unfortunate - why is this needed? Can we avoid this?

File: docs/api.md [hunk 2]
```diff
@@ -1045,9 +1045,9 @@ Examples:
 const divsCounts = await page.$$eval('div', divs => divs.length);
 ```
 
-#### page.$eval(selector, pageFunction[, ...args])
+#### page.$eval(selector, asdf[, ...args])
```
Comment (aslushnikov) [inline docs/api.md]: asdf!

**Review by aslushnikov — COMMENTED** [review]
first pass over this wasn't scary!

File: docs/api.md (hunk 1 above)
Comment (JoelEinbinder) [reply docs/api.md:1038]: Without it, markdown can think the function is a link. It tends not to because of syntax errors, but I'd rather not depend on it bailing out on its own.

File: docs/api.md (hunk 2 above)
Comment (JoelEinbinder) [reply docs/api.md]: oops

**Comment by SimonSchick** [comment]
Can I ask why you choose to not use TS directly?

**Comment by JoelEinbinder** [comment]
For this PR or for the project in general?

**Comment by SimonSchick** [comment]
The whole project, you seem to be using TS specifying type hinting in JSDoc anyways I think using TS all the way would be a better solution.

**Comment by aslushnikov** [comment]
> The whole project, you seem to be using TS specifying type hinting in JSDoc anyways I think using TS all the way would be a better solution.

@SimonSchick we like our current approach: we have no build step and all the benefits of first-class platform citizen (e.g. natural debugging) while having type checks.

**Comment by JoelEinbinder** [comment]
I setup tests for the types that run against typescript@2.1.

File: utils/doclint/check_public_api/test/md-builder-common/result.txt [hunk 3]
```diff
@@ -5,6 +5,9 @@
       "members": [
         {
           "name": "frame",
+          "type": {
+            "name": "[Frame]"
```
Comment (aslushnikov) [inline utils/doclint/check_public_api/test/md-builder-common/result.txt:9]: Do you really want to have "Frame" in brackets?

File: utils/doclint/generate_types/index.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+const fs = require('fs');
+const objectDefinitions = [];
+(async function() {
+  const browser = await puppeteer.launch();
+  const page = (await browser.pages())[0];
+  const api = await Source.readFile(path.join(PROJECT_DIR, 'docs', 'api.md'));
+  const {documentation} = await require('../check_public_api/MDBuilder')(page, [api]);
+  await browser.close();
+  const classes = documentation.classesArray.slice(1);
+  const root = documentation.classesArray[0];
```
Comment (aslushnikov) [inline utils/doclint/generate_types/index.js:14]: puppeteerClass?

File: utils/doclint/generate_types/index.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  const output = `// This file is generated by ${__filename.substring(path.join(__dirname, '..', '..').length)}
```
Comment (aslushnikov) [inline utils/doclint/generate_types/index.js:15]: relativeTo?

File: utils/doclint/check_public_api/Documentation.js [hunk 6]
```diff
@@ -59,14 +63,17 @@ Documentation.Member = class {
   /**
    * @param {string} kind
    * @param {string} name
-   * @param {!Documentation.Type} type
+   * @param {?Documentation.Type} type
    * @param {!Array<!Documentation.Member>} argsArray
    */
-  constructor(kind, name, type, argsArray) {
+  constructor(kind, name, type, argsArray, comment = '', returnComment = '', required = true) {
```
Comment (aslushnikov) [inline utils/doclint/check_public_api/Documentation.js:69]: can we have returnComment the same way we have arguments?

Comment (JoelEinbinder) [reply utils/doclint/check_public_api/Documentation.js:69]: while this makes the return code nicer, it makes the .type code uglier.

File: utils/doclint/check_public_api/MDBuilder.js [hunk 7]
```diff
@@ -30,57 +30,45 @@ class MDOutline {
...
       }
       return {classes, errors};
 
+      /**
+       * @param {HTMLLIElement} element
+       */
       function parseProperty(element) {
-        const str = element.textContent;
-        const name = str.substring(0, str.indexOf('<')).trim();
+        // const firstLine = extractSiblingsIntoFragment(element.firstChild, element.querySelector(':scope > ul'));
```
Comment (aslushnikov) [inline utils/doclint/check_public_api/MDBuilder.js]: remove

**Review by aslushnikov — APPROVED** [review]
